SCP updates : connect session message

matt335672 opened the pull request:
This PR reworks the SCP protocol to split the 'create session' message into separate 'create session' and 'connect session' messages.

This has little user-visible changes, apart from:-
1)  `SessionSockdirGroup` in `sesman.ini`. For v0.10.2 and later, this is generally set to `xrdp` to allow the xrdp process running as `xrdp:xrdp` to connect directly to sockets. With this PR, this is no longer necessary, as sesman makes socket connections on behalf of xrdp. The value can now be set (or defaulted) to root for more secure configurations.
2) Running `xrdp-sesrun` for an already existing session now no longer causes a phantom run of the reconnectwm.sh script.

The point of the PR is to make a connection to a session a separate, identifiable action. This will allow us to implement the following functionality:-
1) A single connection to a session will be enforced. This fixes a number of PRs around chansrv not working on a reconnection.
2) We can provide dynamic information on the connected client (if any)
3) [Later] reconnection can be implemented (See #1443)

I've updated some wiki pages. New copies are on my wiki as follows:-

- https://github.com/matt335672/xrdp/wiki/The-socketdir-directory
- https://github.com/matt335672/xrdp/wiki/SCP-Sequence-Diagrams
- https://github.com/matt335672/xrdp/wiki/Session-management-architecture

At the source level, the following are worthy of note:-
1) A new module `set_int` which allows sets of integers to be manipulated with a bitmap. There's a test suite for this.
2) the `mod_connect()` call now allows a file descriptor to be passed in.  If this is done, the connect can be made immediately by assigning the file descriptor to the transport.
3) An extra state is added to the connection state machine in xrdp_mm.c to allow file descriptors to be acquired from sesman.

I'm making this draft for now while I perform some more testing. Please feel free to ask me any questions about the rationale, the design or the code.

matt335672 committed:
Update SCP with connect session calls

matt335672 committed:
Rename the pre-session list to the SCP list

The name pre-session list makes no sense now, as we need items
to remain on the list after starting the session and before
connecting.

matt335672 committed:
Add macro GUID_ARE_EQUAL for comparing GUIDs

matt335672 committed:
Rework the start session method in SCP

The start session method is reworked to pass a response back to the
client.
If the method was successful, a session list entry is created. This
is different to the provious behaviour, where we created the
session list entry unconditionally.

This new arrangement means that we need a different way to avoid
a race condition where two users may try to create a session at the
same time, and end up with the same display. We do this by keeping
track of newly allocated displays as part of the SCP connections. When
we allocate a display, the SCP connection displays are also taken
into account.

matt335672 committed:
Add set_int type

This type can be used to store sets of integers. It is intended to
be used to keep track of the display numbers allocated to sessions and
SCP connections.

A test suite for the new type is also added.

matt335672 committed:
Update scp_list.h

matt335672 committed:
Add session connect to libipm sesexec interfaces

matt335672 committed:
Plumb connect_session call into sesman and sesexec

matt335672 committed:
Add optional fd parameter to mod_connect() method

The fd parameter can be used to immediately connect to a display
server, if an fd was obtained from sesman

matt335672 committed:
Plumb file descriptors from sesman into xrdp

If sesman sends file descriptors for the display server and
chansrv, use these in preference to trans_connect() calls.

matt335672 committed:
Update xrdp-sesrun for new SCP interface

The SCP connection is now still open after a create session call,
and needs to be explicitly closed to prevent errors being logged.

matt335672 committed:
Prefer SessionSockdirGroup to be set to 'root'

With recent changes to the SCP interface, the xrdp process no longer
needs read access to the user sockdir when sesman is in use.

tsz8899 commented:
Thank you
Could you please describe on the plans listed below

~~~
SessionSockdirGroup in . For v0.10.2 and later, this is generally set to to allow the xrdp process running as to connect directly to sockets. With this PR, this is no longer necessary, as sesman makes socket connections on behalf of xrdp. The value can now be set (or defaulted) to root for more secure configurations.sesman.inixrdpxrdp:xrdp
~~~
Does it represent best practice to set runtime_user=xrdp and runtime_group=xrdp in xrdp.ini, and set SessionSockdirGroup=root in sesman.ini?
Or is it better to not set these values and simply use the defaults in the future?
Is it still necessary to create the xrdp user and group?



~~~
Running for an already existing session now no longer causes a phantom run of the reconnectwm.sh script.xrdp-sesrun
~~~
In my environment, each time a session is reconnected in the current version, reconnectwm.sh is executed. This script is used to write client IPs, rdp /app: remote application program info, and other forwarding information for xrdp.
Will it be possible to add a switch in the future to force or automatically run reconnectwm.sh?



~~~
We can provide dynamic information on the connected client (if any)
~~~
In the current version, I'm using a rough approach in to deliver information to the user's home directory.xrdp_mm.c (IP, domain, remote app, printer, LDAP,info) 
Will there be a reliable channel for this in the future?

matt335672 commented:
Hi @tsz8899 

As usual, you raise some excellent points and questions - thank you.

> Does it represent best practice to set runtime_user=xrdp and runtime_group=xrdp in xrdp.ini, and set  SessionSockdirGroup=root in sesman.ini?
> Or is it better to not set these values and simply use the defaults in the future?
> Is it still necessary to create the xrdp user and group?

With this update, best practice for a normal configuration will be to set `runtime_user` and `runtime_group` to the username and group of an unprivileged account, and simply not set `SessionSockdirGroup`. Some rarer configurations may still require `SessionSockdirGroup` to be set. The only one I'm aware of is when running chansrv in a VNC session which has not been started by sesman.

You will still need to create the unprivileged account, or use an existing one.

> Will it be possible to add a switch in the future to force or automatically run reconnectwm.sh?

I'll add this - great suggestion.

> In the current version, I'm using a rough approach in to deliver information to the user's home directory.xrdp_mm.c (IP, domain, remote app, printer, LDAP,info)
> Will there be a reliable channel for this in the future?

I'm working on a PR based on this one which will give me a channel between the xrdp process and the sesman process. That should let you query the connected client using `xrdp-sesadmin`. The information available will change as different clients disconnect and reconnect.

matt335672 committed:
Add AlwaysRunReconnect config option

This allows the system administrator to specify whether the
reconnectwm.sh script should only be run on reconnects, or should
be run for all connections to a session.

matt335672 committed:
Fix Coverity warnings

Two problems were found:-
1) A useless test in scp_list.c - testing an unsigned int was >= 0.
2) Flow control issues in scp.c:scp_get_connect_session_response() meant
   that file descriptors could be leaked. A helper function has been
   used to simplify the code.
